src/util/piped_process.cpp

```diff
@@ -0,0 +1,233 @@
+  }
+
+  // Default state
+  process_state = process_statet::NOT_CREATED;
+
+  if(fcntl(pipe_output[0], F_SETFL, O_NONBLOCK) < 0)
+  {
+    throw std::runtime_error("Setting pipe non-blocking failed");
+  }
+
+  // Create a new process for the child that will execute the
+  // command and receive information via pipes.
+  pid = fork();
+  if(pid == 0)
+  {
+    // child process here
+
+    // Close pipes that will be used by the parent so we do
+    // not have our own copies and conflicts.
+    close(pipe_input[1]);
+    close(pipe_output[0]);
+
+    // Duplicate pipes so we have the ones we need.
+    dup2(pipe_input[0], STDIN_FILENO);
+    dup2(pipe_output[1], STDOUT_FILENO);
+    dup2(pipe_output[1], STDERR_FILENO);
+
+    // Create the arguments to execvp from the construction string
+    char **args = string_to_cstr_array(command, " ");
+
+    // Execute the command
+    execvp(args[0], args);
+    // The args variable will be handled by the OS if execvp succeeds, but
+    // if execvp fails then we should free it here (just in case the runtime
+    // error below continues execution.)
+    free(args);
+    // Only reachable if execvp failed
+    throw std::runtime_error(
+      "Launching \"" + command + "\" failed with error: " + strerror(errno));
+  }
+  else
+  {
+    // parent process here
+    // Close pipes to be used by the child process
+    close(pipe_input[0]);
+    close(pipe_output[1]);
+
+    // Get stream for sending to the child process
+    command_stream = fdopen(pipe_input[1], "w");
+    process_state = process_statet::CREATED;
+  }
+#endif
+}
+
+piped_processt::~piped_processt()
+{
+#ifdef _WIN32
+  UNIMPLEMENTED_FEATURE("Pipe IPC on windows: piped_processt constructor")
+#else
+  // Close the parent side of the remaning pipes
+  fclose(command_stream);
+  // Note that the above will call close(pipe_input[1]);
+  close(pipe_output[0]);
+  // Send signal to the child process to terminate
+  kill(pid, SIGTERM);
+#endif
+}
+
+piped_processt::process_send_responset
+piped_processt::send(const std::string &message)
+{
+#ifdef _WIN32
+  UNIMPLEMENTED_FEATURE("Pipe IPC on windows: send()")
+#else
+
+  if(process_state != process_statet::CREATED)
+  {
+    return process_send_responset::ERROR;
+  }
+
+  // send message to solver process
+  int send_status = fputs(message.c_str(), command_stream);
+  fflush(command_stream);
+
+  if(send_status == EOF)
+  {
+    // Some kind of error occured, maybe we should update the
+    // process status here?
+    return process_send_responset::FAILED;
+  }
+
+  return process_send_responset::SUCCEEDED;
+#endif
+}
+
+std::string piped_processt::receive()
+{
+#ifdef _WIN32
+  UNIMPLEMENTED_FEATURE("Pipe IPC on windows: receive()")
+#else
+
+  INVARIANT(
+    process_state == process_statet::CREATED,
+    "Can only receive() from a fully initialised process");
+
+  // This is necessary to ensure the buffer is synced with the
+  // latest data.
+  fsync(pipe_output[0]);
```

Comment:
Rather than using explicit syncs/flushes, is it possible to put the pipe into non-buffered mode? There should be an ioctl operation on the pipe for that, I think....

Reply:
We deleted `fsync` as we found out we don't really need it.